tools/istio-iptables/pkg/cmd/run.go

```diff
@@ -357,6 +352,21 @@ func run(config *config.Config) {
 
 	logConfig(config)
 
+	// check if ISTIO_REDIRECT already exists
+	// if it exists, exit, since the chains are already set
+	// the pattern to check if a chain exists, is described at https://stackoverflow.com/a/10784612/553720
+
+	err = ext.Run(dep.IPTABLES, "-t", constants.NAT, "--list", constants.ISTIOREDIRECT)
+	if err == nil {
+		fmt.Println(constants.ISTIOREDIRECT + " detected, exiting")
+		return
```

Comment:
Exit with error.
You don't know whether the chains are correctly configured, i.e. whether all the chains have been created and contain the expected list of rules.
So to be safe please just error out, and let `kubelet` restart the pod and reset the network namespace.

Reply:
Hmmm this runs in an init container, so exiting with an error here will make kubelet continuously restart this init container, without restarting the pod. That's not good. :/

Reply:
So, to clarify the situation:

The init container should run once so the previous Istio iptables chains should not exist. However, due to some bug in the environment, as mentioned in https://github.com/istio/istio/issues/16768, the init container can mistakenly run multiple times. Three possible options to cope with the problem:

1. Clean the previous iptables chains if they exist and create the chains a new. (The previous state, before https://github.com/istio/istio/pull/17044). In this case, if the restart happens in the middle of Istio running, during the cleanup and recreation of the iptables chains, some requests will slip into this window between cleanup and recreation, bypassing the sidecar and causing errors. This is what one of our customer experiences on OpenShift.

1. Fail if the previous iptables chains exist. This will enter the init container in continuous restarting and failing. (The current state, after https://github.com/istio/istio/pull/17044)

1. Check if the previous iptables exist and ignore them silently. This will make Istio resilient to the errors in the environment. The previous iptables should be the same as the ones the script intended to create since the script by design should run only once. There is no such scenario, as far as I know, as creating iptables chains, and then changing them. Iptables chains are created only once per pod. (The proposed state in this PR)

The last option will make Istio resilient, but will mask the possible problems in the environment. I am advocating this approach in https://github.com/istio/istio/issues/16768. Let's make Istio resilient and let the Operations people discover the problems in the environment by means other than Istio failing. It is none of Istio business to discover problems in the user's environments, Istio should be as resilient to the problems as possible.

@rlenglet @howardjohn what do you think?

Reply:
I'm in favor of 3., but this is not what this PR does. It only checks that *one* chain exists. It is possible that the chains are only partially configured, because the previous script run failed before completely configuring all chains and rules.

I would feel more comfortable if the check was more thorough, e.g. it checks that all expected chains are created, and maybe even the redirects from INPUT / OUTPUT are there, etc.

Reply:
@rlenglet sounds good, let me implement it.

Reply:
@rlenglet I think the most thorough way to implement the check is to use [iptables-save](https://linux.die.net/man/8/iptables-save) and [iptables-restore](http://man7.org/linux/man-pages/man8/iptables-restore.8.html).

We can do the following implementation:
1.  Change the script to write the iptables commands into a text file, in the format of iptables-save/iptables-restore, instead of actually executing them. This is similar to operation of some databases that first write their operations to a log file and then actually perform the operations.
2. Once finished writing the iptables commands into a file, the script will dump the current state of iptable chains into another file by `iptables-save`.
3. The script will diff the text files. If the files are identical, the script will exit. If the files are different, the script will run `clean` and `iptables-restore` on the text file it wrote, this way actually creating the iptables chains.

What do you think? I think there are advantages in this mode of operation: `iptables-restore` will check the syntax of the commands and will execute them at once. It is like a transaction, either all the commands success or the whole update (by `iptables-restore`) fails.

Reply:
Please review @abhide's PR: https://github.com/istio/istio/pull/18622. It's doing exactly that, but in the golang implementation. :)

Reply:
@rlenglet Great, will wait for #18622 and then implement the diff/clean/restore/ functionality.

Reply:
I am using `--no-flush` for `iptables-restore`. that way old chains aren't deleted
https://github.com/istio/istio/pull/18622/files#diff-7737d52cc5c9db7a81451826f15ec97eR475.

I like the transaction semantics for iptables-restore; less confusing over `|| true` approach that might mask certain failures.

Reply:
Just FYI, @vadimeisenbergibm https://github.com/istio/istio/pull/18622 is merged

Reply:
@abhide thanks, I will resume my work on this PR taking #18622 into account.